subprojects/docs/src/docs/userguide/gradle_wrapper.adoc

```diff
@@ -238,3 +238,76 @@ distributionSha256Sum=371cb9fbebbe9880d147f59bab36d61eee122854ef8c9ee1ecf12b8236
+Here are the commands you can run on the major operating systems to manually verify a Wrapper JAR's checksum:
+
+[source,bash,subs=+attributes]
+.Manually verifying the checksum of the Wrapper JAR on Linux
+----
+$ cd gradle/wrapper
+$ curl --location --output gradle-wrapper.jar.sha256 \
+       https://services.gradle.org/distributions/gradle-{gradleVersion}-wrapper.jar.sha256
+$ echo "  gradle-wrapper.jar" >> gradle-wrapper.jar.sha256
+$ sha256sum --check gradle-wrapper.jar.sha256
+gradle-wrapper.jar: OK
+----
+
+[source,bash,subs=+attributes]
+.Manually verifying the checksum of the Wrapper JAR on macOS
+----
+$ cd gradle/wrapper
+$ curl --location --output gradle-wrapper.jar.sha256 \
+       https://services.gradle.org/distributions/gradle-{gradleVersion}-wrapper.jar.sha256
+$ echo "  gradle-wrapper.jar" >> gradle-wrapper.jar.sha256
+$ shasum --check gradle-wrapper.jar.sha256
+gradle-wrapper.jar: OK
+----
+
+[source,powershell,subs=+attributes]
+.Manually verifying the checksum of the Wrapper JAR on Windows (using PowerShell)
+----
+> $expected = Invoke-RestMethod -Uri https://services.gradle.org/distributions/gradle-{gradleVersion}-wrapper.jar.sha256
+> $actual = (Get-FileHash gradle-wrapper.jar -Algorithm SHA256).Hash.ToLower()
+> @{$true = 'OK: Checksum match'; $false = "ERROR: Checksum mismatch!`nExpected: $expected`nActual:   $actual"}[$actual -eq $expected]
+OK: Checksum match
+----
+
+=== Troubleshooting a checksum mismatch
+
+If the checksum does not match the one you expected, chances are the `wrapper` task wasn't executed with the upgraded Gradle distribution.
+Thus, you should first check whether the actual checksum matches the one of a different Gradle version.
+Here are the commands you can run on the major operating systems to generate the actual checksum of the Wrapper JAR:
+
+[source,bash,subs=+attributes]
+.Generating the actual checksum of the Wrapper JAR on Linux
+----
+$ sha256sum gradle/wrapper/gradle-wrapper.jar
+d81e0f23ade952b35e55333dd5f1821585e887c6d24305aeea2fbc8dad564b95  gradle/wrapper/gradle-wrapper.jar
+----
+
+[source,bash,subs=+attributes]
+.Generating the actual checksum of the Wrapper JAR on macOS
+----
+$ shasum --algorithm=256 gradle/wrapper/gradle-wrapper.jar
+d81e0f23ade952b35e55333dd5f1821585e887c6d24305aeea2fbc8dad564b95  gradle/wrapper/gradle-wrapper.jar
+----
+
+[source,powershell,subs=+attributes]
+.Generating the actual checksum of the Wrapper JAR on Windows (using PowerShell)
+----
+> (Get-FileHash gradle\wrapper\gradle-wrapper.jar -Algorithm SHA256).Hash.ToLower()
+d81e0f23ade952b35e55333dd5f1821585e887c6d24305aeea2fbc8dad564b95
+----
+
+Once you know the actual checksum, check whether it's listed on https://gradle.org/wrapper-checksums/.
+If so, you should run the `wrapper` task again to update the Wrapper JAR so it matches the Gradle version in `gradle/wrapper/gradle-wrapper.properties`.
```

Comment:
:x: I would **not** advise the user to run the `wrapper` task if the checksum doesn't match. This needs to be rephrased somewhat.
What it's implying is that the user should update the wrapper with the potentially malicious wrapper.

Reply:
@JLLeitschuh, you got this wrong, this advise the user to run `:wrapper` when the wrapper jar checksum matches a different Gradle version than the one currently in use. That's the "run `:wrapper` task twice" problem.

@marcphilipp, this sentence could be made more explicit so people glancing at this guide don't take that sentence as Jonathan did.